### DISCUSSION TITLE
Cannot access iModel/Reality Data once created

### DISCUSSION THREAD

**Discussion by bighamj** [iTwin/itwinjs-core#4743] [Q&A] [answered]
Hello all! 

I have created a service which:
1. Creates an iModel
2. Creates Reality Data attached to that iModel
3. Uploads tiled scan data to that Reality Data.

My issue:
- When I try to query for the iModel, I get 404 (not found) when using the documentation website with the same account that I created the service app.
- If I query my permissions with my account, I get the fancy admin-type permissions, itwin_read, etc. However, if I try to query my permission for a created iModel created by the service owned by that account, I get empty permissions.
- For reference, I'm using this page to query my permissions: https://developer.bentley.com/apis/access-control/operations/get-iTwin-permissions/

Thanks for your help in advance!

**Comment by maksim-bnt** [comment, marked as answer]
Hi,

Can you please provide more information on how you created iModel ? Did you create iModel with a service app's token and then tried to access it with your token?

**Reply by bighamj** [reply]
Hello!

That’s correct. I created it with the imodel api and a service token and want to access it in a front end react app.

**Reply by maksim-bnt** [reply]
Because iTwin and iModel are created with a service token you also need to grant access to your personal identity on that iTwin. To grant access make the following API calls with the service token.

1. Create iTwin Role:  https://developer.bentley.com/apis/access-control/operations/create-iTwin-role/
2. Assign iModel permissions: https://developer.bentley.com/apis/access-control/operations/update-iTwin-role/ . The permissions will depend on what operations front end should be able to do. For example for read-only assign 'imodels_read', to modify assign 'imodels_write'. 
3. Assign the role created in (1) to your identity: https://developer.bentley.com/apis/access-control/operations/add-iTwin-members/

**Reply by bighamj** [reply]
Hello there again! 

I was able to accomplish what you outlined. Thanks for taking the time to provide the steps!

Now I'm trying to view those within the viewer, but I'm not understanding how to authenticate for this. 

I see that @itwin/web-viewer-react requires a ViewerAuthorizationClient to be implemented. Is there an example of how to do this without requiring user sign in? Similar to a backend service?

Thanks again for your help!

**Reply by MichaelBelousov** [reply]
@bighamj let me try my best to answer here and someone else can come and correct me when they're available.

You must log in to view a standard iModel, because all iModels have associated users and permissions. I do not believe there is a secure way to access the necessary credentials to just view the iModel. The analogy of a backend service's login flow doesn't apply securely because you run the backend on a private machine, where as people opening your viewer site would need those credentials to do it the same way. I will try to see what iTwin sandbox does to share a viewer of the sample iModels since that is somewhat similar.

**Reply by MichaelBelousov** [reply]
@bighamj can you log in and see the iModel here? https://mytwin-viewer.netlify.app/

The creator of that template is on vacation, but I need to confirm if there's something special about the sample iModels that makes them viewable with publicly, and if there is, I'm not sure it's something you can do yourself.

**Reply by aruniverse** [reply]
Hi @bighamj ,

> I see that @itwin/web-viewer-react requires a ViewerAuthorizationClient to be implemented. Is there an example of how to do this without requiring user sign in? Similar to a backend service?

Please refer to https://github.com/calebmshafer/itwin-viewer-proxy which shows an example of how to solve this.

**Comment by bighamj** [comment]
Hello everyone, thanks for all the answers! 

@aruniverse I think I found that exact repo in my keyboard smashing! That's how I got it to work, I have an internal API that can generate the token.

Now to move on to struggles with the Reality Data API! (Had a struggle with file formats when using Cesium API to convert then upload to Reality Data API). 

![image](https://user-images.githubusercontent.com/66083892/210433877-75641273-4ca9-4921-8187-61e619d534dc.png)
